### DISCUSSION TITLE
Shell/property injection: know when it's finished

### DISCUSSION THREAD

**Discussion by softlion** [dotnet/maui#4772] [Ideas] [answered]
Hello mauiers,
it would be nice to have a method called when property injection has finished. So custom initialization can start.

Something like:

```c#
[QueryProperty(nameof(Item1), "item1")]
[QueryProperty(nameof(Item2), "item2")]
[QueryInitialized(nameof(MyMethod))]
public class MyModel
{
    public string? Item1 { get; set; }
    public string? Item2 { get; set; }

    private void MyMethod()
    {
        //The object knows that Item1 and Item2 have been initialized.
    }
}

```

What do you think ?

The current workaround is to forward `OnParentSet` from the view. But this has a side effect of displaying an uninitialized view that may flash to its initialized state.

```c#
protected override void OnParentSet()
{
    base.OnParentSet();
    ((MyModel)BindingContext).MyMethod();
}
```

**Comment by Jared-Aus** [comment, marked as answer]
I'm not a fan of shell navigation due to complexities like these, but I think in your case [IQueryAttributable](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/app-fundamentals/shell/navigation#process-navigation-data-using-a-single-method) might be a reasonable substitute. [](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/app-fundamentals/shell/navigation#process-navigation-data-using-a-single-method) It won't solve the likely problem of your view model being in a useless/empty state when the page is first presented however.

**Reply by softlion** [reply]
Ok that won't resolve my issue, but you can't know :)

I'm reusing a viewmodel between 2 pages. One page is accessed by a tab. So it can't have any query param at all, as tab pages don't have that.
The other page is a normal child page that does have ome query params.

My issue was to have a common entry point in the viewmodel that is triggered after the optional query params have been set, or not set (ie: the tab is displayed for the 1st time).

But that's a good point !
